File: capstoneproject/messaging_system/services/reaction_service.py
"""
services/reaction_service.py
Handles reaction operations for messages in the messaging system.
"""

class ReactionService:
    """Manages reactions to messages."""
    
    def __init__(self):
        """Initialize the reaction service."""
        pass

    def add_reaction(self, message_id: int, user_id: int, reaction_type: str) -> bool:
        """Add a reaction to a message.
        
        Args:
            message_id: ID of the message to react to
            user_id: ID of the user adding the reaction
            reaction_type: Type of reaction (like, dislike, etc.)
            
        Returns:
            True if reaction was added successfully
        """
        pass

    def remove_reaction(self, message_id: int, user_id: int) -> bool:
        """Remove a user's reaction from a message.
        
        Args:
            message_id: ID of the message
            user_id: ID of the user removing reaction
            
        Returns:
            True if reaction was removed successfully
        """
        pass

    def get_reactions(self, message_id: int) -> list:
        """Get all reactions for a specific message.
        
        Args:
            message_id: ID of the message
            
        Returns:
            List of reactions for the message
        """
        pass

File: capstoneproject/app/message/services/ReactionService.ts
/**
 * services/ReactionService.ts
 * Handles reaction operations for messages in the messaging system.
 */

import type { ReactionData, ReactionType } from '../models/Reaction';
import { Reaction } from '../models/Reaction';

export interface AddReactionRequest {
  messageId: number;
  userId: number;
  type: ReactionType;
}

export class ReactionService {
  /**
   * Manages reactions to messages.
   */
  private baseUrl: string;

  constructor(baseUrl: string = '/api/reactions') {
    this.baseUrl = baseUrl;
  }

  /**
   * Add a reaction to a message.
   */
  async addReaction(reactionData: AddReactionRequest): Promise<Reaction> {
    const response = await fetch(this.baseUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(reactionData)
    });

    if (!response.ok) {
      throw new Error('Failed to add reaction');
    }

    const data: ReactionData = await response.json();
    return new Reaction(data);
  }

  /**
   * Remove a user's reaction from a message.
   */
  async removeReaction(messageId: number, userId: number): Promise<boolean> {
    const response = await fetch(`${this.baseUrl}/${messageId}/user/${userId}`, {
      method: 'DELETE'
    });

    return response.ok;
  }

  /**
   * Get all reactions for a specific message.
   */
  async getReactions(messageId: number): Promise<Reaction[]> {
    const response = await fetch(`${this.baseUrl}/message/${messageId}`);
    
    if (!response.ok) {
      throw new Error('Failed to fetch reactions');
    }

    const data: ReactionData[] = await response.json();
    return data.map(reactionData => new Reaction(reactionData));
  }

  /**
   * Get reaction counts grouped by type for a message.
   */
  async getReactionCounts(messageId: number): Promise<Record<ReactionType, number>> {
    const reactions = await this.getReactions(messageId);
    const counts: Record<ReactionType, number> = {} as Record<ReactionType, number>;

    reactions.forEach(reaction => {
      counts[reaction.type] = (counts[reaction.type] || 0) + 1;
    });

    return counts;
  }

  /**
   * Toggle user's reaction on a message (like ↔ dislike)
   */
  async toggleReaction(messageId: number, userId: number, newType: ReactionType): Promise<Reaction> {
    // Remove existing reaction first, then add new one
    await this.removeReaction(messageId, userId);
    return await this.addReaction({ messageId, userId, type: newType });
  }
}
